samcli/local/services/localhost_runner.py

```diff
@@ -0,0 +1,124 @@
+import os
+
+from flask import Response
+
+LOG = logging.getLogger(__name__)
+
+
+class LocalhostRunner(object):
+
+    def __init__(self, lambda_runner, port, host, stderr=None):
+        """
+        Creates a BaseService class
+
+        Parameters
+        ----------
+        lambda_runner samcli.commands.local.lib.local_lambda.LocalLambdaRunner
+            The Lambda runner class capable of invoking the function
+        port int
+            Optional. port for the service to start listening on Defaults to 3000
+        host str
+            Optional. host to start the service on Defaults to '127.0.0.1
+        stderr io.BaseIO
+            Optional stream where the stderr from Docker container should be written to
+        """
+        self.lambda_runner = lambda_runner
+        self.port = port
+        self.host = host
+        self._app = None
+        self.stderr = stderr
+
+    def create(self):
+        """
+        Creates a Flask Application that can be started.
+        """
+        raise NotImplementedError("Required method to implement")
+
+    def run(self):
+        """
+        This starts up the (threaded) Local Server.
+        Note: This is a **blocking call**
+
+        Raises
+        ------
+        RuntimeError
+            if the service was not created
+        """
+        if not self._app:
+            raise RuntimeError("The application must be created before running")
+
+        # Flask can operate as a single threaded server (which is default) and a multi-threaded server which is
+        # more for development. When the Lambda container is going to be debugged, then it does not make sense
+        # to turn on multi-threading because customers can realistically attach only one container at a time to
+        # the debugger. Keeping this single threaded also enables the Lambda Runner to handle Ctrl+C in order to
+        # kill the container gracefully (Ctrl+C can be handled only by the main thread)
+        multi_threaded = not self.lambda_runner.is_debugging()
+
+        LOG.debug("Localhost server is starting up. Multi-threading = %s", multi_threaded)
+
+        # This environ signifies we are running a main function for Flask. This is true, since we are using it within
+        # our cli and not on a production server.
+        os.environ['WERKZEUG_RUN_MAIN'] = 'true'
+
+        self._app.run(threaded=multi_threaded, host=self.host, port=self.port)
+
+    @staticmethod
+    def _service_response(body, headers, status_code):
+        """
+        Constructs a Flask Response from the body, headers, and status_code.
+
+        :param str body: Response body as a string
+        :param dict headers: headers for the response
+        :param int status_code: status_code for response
+        :return: Flask Response
+        """
+        response = Response(body)
+        response.headers = headers
+        response.status_code = status_code
+        return response
+
+
+class LambdaOutputParser(object):
```

Comment:
optional: doesn't need a class. It could be just its own top-level method

Reply:
I am going to leave this as a class for now. We need to read the output for the Lambda to see if the return value from the container is a user exception/error. This is so we can properly handle some error cases and setting headers correctly. I was planning to use this class to contain that as well. I may just move it then but want to see how the structure/code looks after I handle that.